### PR TITLE
Update EIP-3540: Move to Review

### PR DESCRIPTION
**Pull request by pdobacz**
First step to update to be in sync with [Megaspec](https://github.com/ipsilon/eof/blob/main/spec/eof.md).

Opening as draft first to get a round of reviews on these changes, then update at least the EIP-7480 and add the CREATE3/CREATE4 EIP to keep references working. (EDIT: both done, CREATE3/4 EIP is EIP-7620)

Let's focus the first round of reviews on overall structure and scope of this doc, and also any potential issues with the spec itself, which may come to mind when looking through these changes -- if anything looks suspicious here please report.

Note that many of the removed portions concern contract creation which ~will be described from ground up in a new EIP (current specification is inside the [Megaspec](https://github.com/ipsilon/eof/blob/main/spec/eof.md)~ is EIP-7620).

### CODE REVIEW

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:115]: Linter requires to change underscore to asterisk here.

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:60]: I think it's fine to leave like before, in you version "EOF" is used before being defined in next sentence.
```suggestion
We introduce *code validation* for new contract creation. To achieve this, we define a format called EVM Object Format (EOF), containing a version indicator, and a ruleset of validity tied to a given version.
```

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:64]:
```suggestion
Code validation is performed during `CREATE4` instruction, and is elaborated on in [EIP-3670](./eip-3670.md) and an upcoming contract creation EIP-???? (TBD, refer to the [Megaspec doc](https://github.com/ipsilon/eof/blob/main/spec/eof.md) until then)
```

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:83]: Not sure why this part was changed, I think previous general description still applies?

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:115]: Add also that square brackets mean optionality (0 or 1).

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:126]:
```suggestion
| num_code_sections      | 2 bytes  | 0x0001-0x0400 | 16-bit unsigned big-endian integer denoting the number of the code sections             |
```

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:135]:
```suggestion
(*) For not yet deployed containers this can be greater than the actual content length, see Data Section Lifecycle section in the contract creation EIP-???? (TBD + link).
```

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:144]:
```suggestion
| max_stack_height  | 2 bytes  | 0x0000-0x03FF | maximum number of elements ever placed onto the operand stack by the code section             |
```

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:146]:
```suggestion
| code_section      | variable | n/a           | arbitrary bytecode                                                         |
| container_section | variable | n/a           | arbitrary EOF-formatted container                                                           |
```

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:166]: I think these are not rules and should be removed

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:163]: I would also add
```
- data section is mandatory, but `data_size` may be 0
```

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:146]: This part was mandating exact order of the sections, we are kind of missing it now, but maybe the formula in `Container` section is enough.

File: EIPS/eip-3540.md
Comment (gumb0) [reply EIPS/eip-3540.md:163]: Also maybe something like
```
- data body length may be shorter than `data_size`
```

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:171]: Linter requirement
```suggestion

- Execution starts at the first byte of code section 0, and `pc` is set to 0.
```

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:173]: I think we should either remove this, or have a full list of deprecated instructions. But it should be in EIP-3670, so maybe remove.

File: EIPS/eip-3540.md
Comment (pdobacz) [reply EIPS/eip-3540.md:83]: It doesn't - for code and containers sections it is not a 2-byte size anymore.

File: EIPS/eip-3540.md
Comment (pdobacz) [reply EIPS/eip-3540.md:123]: got it from Megaspec. In general, once dust settles here, I'll port the refreshed format of the tables back to there :)

File: EIPS/eip-3540.md
Comment (pdobacz) [reply EIPS/eip-3540.md:124]: Makes sense, but this is taken from the megaspec, so the values there incorrect? Are we 100% sure the change wasn't intended?

File: EIPS/eip-3540.md
Comment (pdobacz) [reply EIPS/eip-3540.md:146]: yea, I assumed it's enough and much cleaner than this

File: EIPS/eip-3540.md
Comment (pdobacz) [reply EIPS/eip-3540.md:173]: After some deliberation, I decided to have the deprecations accompany the replacement definitions wherever possible, and here list all the deprecations without replacements. The EIPs also kinda followed this model, so retaining this should give a smaller diff. But mostly, I think it reads easier, if the legacy and EOF logical counterparts are nearby.

But that's not a strong opinion, I spent quite a lot of time juggling these alternatives, all have pros and cons

File: EIPS/eip-3540.md
Comment (gumb0) [reply EIPS/eip-3540.md:83]: `section_size` applied only to single-size sections (type and data), and `section_size_list` applied to code sections and now should also apply to container sections. The paragraph above described this well:

> The EOF header is followed by at least one section header. Each section header contains two fields, `section_kind` and either `section_size` or `section_size_list`, depending on the kind. `section_size_list` is a list of size values when multiple sections of this kind are allowed.

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:106]:
```suggestion
    kind_type, type_size, 
```

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:164]:
```suggestion
- the number of code sections must not exceed `1024`
- `code_size` may not be `0`
- the number of container sections must not exceed `256`
- `container_size` may not be `0`, but container sections are optional
- data section is mandatory, but `data_size` may be `0`
- data body length may be shorter than `data_size` for a not yet deployed container
```

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:233]: Why did you make PC lowercase?

File: EIPS/eip-3540.md
Comment (pdobacz) [reply EIPS/eip-3540.md:83]: Oh, so `_list` here included the number of items as the first `uint16` :facepalm: . I somehow misread this and the "or", and it seemed completely out-of-date

Nevertheless. I thought we wanted to make the general EOF header format more liberally defined, per the rationale included below:

> - Whether or not to have more structure to the container header for all EOF versions to follow. In order to allow future formats optimized for chunking and merkleization (verkleization) it was decided to keep it generic and specify the structure only for a specific EOF version.

This would invite putting offsets in, rather than sizes.

Should we revert this and require a specific structure for all versions of EOF?

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:249]: I would just remove Test Cases section and focus on providing state tests instead.

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:258]: I think we decided that new calls are only for EOF?
```suggestion
- EOF0 `DELEGATECALL` to EOF1 code succeeds
```

File: EIPS/eip-3540.md
Comment (pdobacz) [reply EIPS/eip-3540.md:255]: yup, leftovers again :(. But I'm removing test cases anyway

File: EIPS/eip-3540.md
Comment (pdobacz) [reply EIPS/eip-3540.md:258]: correct, I updated too much here

File: EIPS/eip-3540.md
Comment (pdobacz) [reply EIPS/eip-3540.md]: Agreed on the call to revert, https://github.com/ethereum/EIPs/pull/8152/commits/7e4c93bfc507dd84a59cb6dfe02f8a6711e84258 (I also took the liberty to be more specific about how the list is encoded)

**Comment by chfast** [comment]
> The commit [7e4c93b](https://github.com/ethereum/EIPs/commit/7e4c93bfc507dd84a59cb6dfe02f8a6711e84258) (as a parent of [8fde16c](https://github.com/ethereum/EIPs/commit/8fde16cdbaa8ad03550d02d0c5beb13dd7838d63)) contains errors. Please inspect the [Run Summary](https://github.com/ethereum/EIPs/pull/8152/files) for details.

This is blocked by missing EIP-7620 file.

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:232]: I guess
```suggestion
Moved to the section [Lack of `EXTDATACOPY` in EIP-7480](./eip-7480.md#lack-of-extdatacopy).
```

**Comment by SamWilsn** [comment]
Just FYI, this'll need to wait for 7620 to get into Review before it can merge.

File: EIPS/eip-3540.md
Comment (axic) [inline EIPS/eip-3540.md:11]: I'd remove the circular dependencies: 4750, 5450, 7620.

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:159]: It's mentioned in the table, but maybe worth to mention here explicitly that `num_code_sections > 0`.

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:161]: Maybe worth to mention here explicitly that `num_container_sections > 0`, if they are present (you can't have container section declared with 0 number of containers)

File: EIPS/eip-3540.md
Comment (gumb0) [reply EIPS/eip-3540.md]: I'm not a fan of this, because without it it seems you can implement the format according to EIP and then pass 0 tests.

Maybe I'm wrong, because you wouldn't pass anything without implementing the rest of validation anyway.

Comment (pdobacz) [reply EIPS/eip-3540.md]: Following the convention we seem to have adopted recently with the other EIPs, I'll leave a note that one of `outputs`' special values has already been assigned in a separate EIPs, or sth like that.

File: EIPS/eip-3540.md
Comment (rodiazet) [inline EIPS/eip-3540.md:167]: This list is different than [Container Validation](https://github.com/ipsilon/eof/blob/main/spec/eof.md#container-validation)

File: EIPS/eip-3540.md
Comment (rodiazet) [inline EIPS/eip-3540.md:175]: Can we add a reference?

File: EIPS/eip-3540.md
Comment (rodiazet) [inline EIPS/eip-3540.md:75]: Can we add a ref to The MAGIC section below?

File: EIPS/eip-3540.md
Comment (rodiazet) [inline EIPS/eip-3540.md:234]: I think we don't have to mentioned that it's moved. We can just add a reference to "Lack of ..."

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:236]: I think this entire paragraph is outdated, it explained why PC opcode returns what it does, when it was still allowed. It is irrelevant now.

File: EIPS/eip-3540.md
Comment (pdobacz) [reply EIPS/eip-3540.md:75]: As mentioned elsewhere, I think this is a tiny bit non-standard to link to the Rationale section bits. If done scrupulously for all the pieces of the spec, would overwhelm the spec itself.

**Review by rodiazet — APPROVED** [review]

**Review by gumb0 — APPROVED** [review]

File: EIPS/eip-3540.md
Comment (gumb0) [inline EIPS/eip-3540.md:173]: Probably this also can be removed.

Comment (pdobacz) [reply EIPS/eip-3540.md:173]: right, but isn't it a remark on how the context variable PC works, rather than opcode? I never was really sure. If you think opcode - let's remove. Actually, I think EIP-4750 elaborates enough on how PC works as a context variable, so I guess we can remove anyway.

Comment (gumb0) [reply EIPS/eip-3540.md]: I thought so too at first, but it's next to the list of what happens to other opcodes, so very much looks like about opcode too.

Comment (pdobacz) [reply EIPS/eip-3540.md]: 088f5e25 a final check on this pls. I think this leaves PC out completely as irrelevant, making 3540 define how EOF would have worked without CALLF/RETF (only 0th code section executes), which I think is coherent.

**Review by gumb0 — APPROVED** [review]

**Pull request merged**
